### PR TITLE
[BEAM-1340,BEAM-1345] Tighten up access and use internal annotations a bit in the Java SDK

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

This is a pass over `sdk.runners`, `sdk.transforms`, `sdk.transforms.reflect` (which is entirely internal), `sdk.transforms.windowing` trigger parts, and `sdk.values`

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @davorbonaci @tgroh

**Comment by kennknowles** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformMatcher.java [hunk 1]
```diff
@@ -20,12 +20,16 @@
 
 import org.apache.beam.sdk.annotations.Experimental;
 import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.annotations.Internal;
 import org.apache.beam.sdk.transforms.AppliedPTransform;
 import org.apache.beam.sdk.transforms.PTransform;
 
 /**
- * Matches applications of {@link PTransform PTransforms}.
+ * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformMatcher.java]: update `package-info.java` to also have this disclaimer

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java [hunk 2]
```diff
@@ -47,9 +48,12 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * Captures information about a collection of transformations and their
+ * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Inner class as well

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/AppliedPTransform.java [hunk 3]
```diff
@@ -20,16 +20,14 @@
...
 import org.apache.beam.sdk.values.POutput;
 import org.apache.beam.sdk.values.PValue;
 import org.apache.beam.sdk.values.TupleTag;
 
 /**
- * Represents the application of a {@link PTransform} to a specific input to produce
- * a specific output.
- *
- * <p>For internal use.
+ * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/AppliedPTransform.java]: Add in "An immutable view of a {@link TransformHierarchy.Node}."

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineFnBase.java [hunk 4]
```diff
@@ -31,14 +32,19 @@
...
+ * <p>This class contains the shared interfaces and abstract classes for different types of combine
  * functions.
  *
  * <p>Users should not implement or extend them directly.
  */
+@Internal
 public class CombineFnBase {
   /**
-   * A {@code GloballyCombineFn<InputT, AccumT, OutputT>} specifies how to combine a
+   * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineFnBase.java]: "Users should extend `CombineFn` directly."

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ViewFn.java [hunk 5]
```diff
@@ -36,6 +39,7 @@
  * @param <PrimitiveViewT> the type of the underlying primitive view, provided by the runner
  *        {@code <ViewT>} the type of the value(s) accessible via this {@link PCollectionView}
  */
+@Internal
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ViewFn.java]: Also `Experimental`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java [hunk 6]
```diff
@@ -19,10 +19,16 @@
...
 import org.joda.time.Duration;
 import org.joda.time.Instant;
 
-/** An abstract description of a standardized transformation on timestamps. */
+/**
+ * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
+ *
+ * <p>An abstract description of a standardized transformation on timestamps.
+ */
+@Internal
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java]: private/package-private constructor also gets you this.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Trigger.java [hunk 7]
```diff
@@ -117,18 +118,26 @@ public Trigger getContinuationTrigger() {
...
   public abstract Instant getWatermarkThatGuaranteesFiring(BoundedWindow window);
 
-  /** Returns whether this performs the same triggering as the given {@link Trigger}. */
+  /**
+   * <b><i>For internal use only; no backwards-compatibility guarantees.</i></b>
+   *
+   * <p>Returns whether this performs the same triggering as the given {@link Trigger}.
+   */
+  @Internal
   public boolean isCompatible(Trigger other) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Trigger.java]: Can this just be replaced with `equals`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java (hunk 6 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/TimestampTransform.java]: I'll do package-private constructor and `@Internal` methods.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ViewFn.java (hunk 5 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ViewFn.java]: I think things that are `@Internal` trump `@Experimental`. It means users just can't use them at all without being broken.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Trigger.java (hunk 7 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Trigger.java]: Unclear. Let's revisit after we tighten up the annotation stuff.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformMatcher.java (hunk 1 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/PTransformMatcher.java]: I almost did, but `PipelineRunner` is public. It should really just be toplevel.

**Comment by kennknowles** [comment]
The only test failure looks like a flake in `HBaseIOTest`.

**Comment by kennknowles** [comment]
retest this please

**Pull request merged**